[Data Views][Grid Layout] Making cmd/ctrl clicking more apparent

andrewhayward opened the pull request:
See also: #59188
Will need to be refactored a bit if #59279 lands first, but the concept will be the same.

## What?

Makes it clearer that <kbd>Cmd</kbd>/<kbd>Ctrl</kbd>-clicking on a grid item is an available action.


## Why?

It is not currently clear that grid items can be <kbd>Cmd</kbd>/<kbd>Ctrl</kbd>-clicked.


## How?

- Changes the grid item cursor when the <kbd>Cmd</kbd>/<kbd>Ctrl</kbd> key is pressed, rather than waiting for the click.

## Testing Instructions

- Moving the mouse pointer over a grid item shouldn't indicate any change
- Pressing and holding the <kbd>Cmd</kbd>/<kbd>Ctrl</kbd> key when the pointer is over a grid item should change the cursor to a `➕`
- Pressing and holding the <kbd>Cmd</kbd>/<kbd>Ctrl</kbd> key then moving the pointer over a grid item should change the cursor to a `➕`
- Pressing and holding the <kbd>Cmd</kbd>/<kbd>Ctrl</kbd> key then clicking anywhere on a grid item should toggle its selected state. 

### Testing Instructions for Keyboard

This change does not impact keyboard interaction.

## Screenshot

![Grid interaction showing cursor/behavioural changes with the command key pressed](https://github.com/WordPress/gutenberg/assets/159848/8ac7a6be-fe08-4762-a9fc-03be422f2624)

Commit:
Making cmd/ctrl clicking more apparent

andrewhayward commented on packages/dataviews/src/view-grid.js:
```diff
@@ -135,6 +131,49 @@ export default function ViewGrid( {
+				eventControllerRef.current = null;
+			} );
+			const config = { capture: true, signal: controller.signal };
+			doc.body.addEventListener( 'keydown', listener, config );
+			doc.body.addEventListener( 'keyup', listener, config );
```
Adding `keyup`/`keydown` listeners when the mouse pointer enters the grid, and removing when the mouse pointer leaves again. The event listeners are added to the body so that the key press is captured regardless of which element has focus.

ntsekouras commented on packages/dataviews/src/view-grid.js:
```diff
@@ -155,16 +194,20 @@ export default function ViewGrid( {
 					alignment="top"
 					className="dataviews-view-grid"
 					aria-busy={ isLoading }
+					onMouseEnter={ pointerIsWithinBoundsHandler }
```
Curious why you added these events in Grid and not in GridItem..

andrewhayward replied:
Mostly just to reduce the number of events being processed; we're manually adding and removing `keyup` and `keydown` event handlers to the body, and doing that every time the mouse pointer moves in and out of a grid item could get expensive.

The `key*` event handlers have to be on the body, because focus could be anywhere; putting them on each grid item would require them to have focus to function properly.

ntsekouras commented on packages/dataviews/src/view-grid.js:
```diff
@@ -44,82 +45,77 @@ function GridItem( {
+				</VStack>
+			</div>
+			{ hasNoPointerEvents && (
+				<div
```
This is  probably here for a11y reasons(role=checkbox), right? If yes, could we move them to the `dataviews-view-grid__content` div?

ntsekouras commented on packages/dataviews/src/style.scss:
```diff
@@ -493,8 +493,23 @@
+			left: 0;
+			right: 0;
+			bottom: 0;
+			cursor: cell;
```
If we go with this design, maybe we should change the cursor to a `-` if it's already selected? 🤔

jameskoster replied:
I don't think such a cursor exists.

andrewhayward replied:
The `cell` cursor is only incidentally a ➕, and as @jameskoster says, there is no ➖ equivalent unfortunately.

ntsekouras commented:
Thanks for the PR @andrewhayward! I agree that this is something we can improve and would love some design feedback about it. Is a cursor enough, or an approach with a light overlay on the grid item or something else..

jameskoster commented:
I don't mind the cursor, it seems to fit the purpose, and is certainly an improvement than the mixture you see currently. 

An alternative (or addition if we want further indication) could be to apply the `is-selected` styles, or maybe just the border, while hovering and pressing cmd/ctrl. If that's trivial to implement then could be worth a try? With hover styles perhaps `cursor: default` would be adequate? Many interactive elements use that cursor in plain html (buttons, checkboxes, radios etc.).

jasmussen commented:
Would you be able to add a visual, or clearer instructions on how to replicate this?

andrewhayward commented:
> An alternative (or addition if we want further indication) could be to apply the `is-selected` styles, or maybe just the border, while hovering and pressing cmd/ctrl.

We can pretty much do whatever we want at this stage. Easy enough to change the border colour, scale the grid item, etc.

![Three cards in a row, the middle one alternating its hover state with a blue border](https://github.com/WordPress/gutenberg/assets/159848/6b11aec2-7780-4d62-99e4-c66404b5ff03)
![Three cards in a row, the middle one alternating its hover state with a floating effect](https://github.com/WordPress/gutenberg/assets/159848/ee5d1816-18c4-46ae-a991-600807107242)

jameskoster commented:
Scaling could be neat but we don't have many other examples, so it may be better to stick with the `is-selected` styles for now.

jasmussen commented:
Thank you for the added visual 🙏 

I actually think the crosshair is not going to be a good fit here. I virtually never see the crosshair cursor used, and seeing it here would actually add to my confusion: am I doing something wrong? It's similar to selecting files in the Finder or Windows Explorer, seeing a crosshair when you select files would likely be equally jarring.

To that end, if we are going to change the cursor, I'd change it to `cursor: default;`. That can keep the `pointer` cursor meaning _click to navigate_.

Commit:
Minor tweaks

jameskoster commented:
As I mentioned above, if we use the `default` cursor, then I think we should also apply the `is-selected` styles to help indicate interactivity.

Commit:
Merge branch 'trunk' into 59188/cmd-ctrl-click

Commit:
Style updates

andrewhayward commented:
Updated the styling, and updated the preview screenshot.

![Grid interaction showing cursor/behavioural changes with the command key pressed](https://github.com/WordPress/gutenberg/assets/159848/8ac7a6be-fe08-4762-a9fc-03be422f2624)

jasmussen commented:
What's the "not allowed" cursor meant to imply?

I genuinely think we should be careful with changing the cursor too much. It feels so jarring when the system cursor changes to something you don't see that often, especially in this case it feels like it's suggesting you did something wrong. I'd personally be most comfortable with using primarily default and pointer cursors, swapping between those two, and then combining with any visual style changes in the list design itself.

andrewhayward commented:
> What's the "not allowed" cursor meant to imply?

To put the card into this "unified checkbox" state, it requires an explicit key press (<kbd>Cmd</kbd>/<kbd>Ctrl</kbd>); the thinking was that it may not be immediately clear to the user why the click didn't register, and whether their key press wasn't picked up. Essentially, we're acknowledging that they might want to take the action, but that they can't.

> I genuinely think we should be careful with changing the cursor too much. It feels so jarring when the system cursor changes to something you don't see that often, especially in this case it feels like it's suggesting you did something wrong.

Definitely agreed that we should be wary of how we use cursors, and I wouldn't have included this here if there was no other interaction involved; the disabled checkboxes for example are fine with a standard cursor. But as mentioned above, because the user has already started the action with the key press, it's helpful to be explicit about the fact it can't be completed in that context.

> I'd personally be most comfortable with using primarily default and pointer cursors, swapping between those two, and then combining with any visual style changes in the list design itself.

All that being said, I don't have a particularly strong opinion, and I'm happy to go with whatever we collectively think is most appropriate. Mostly just exploring options.

jameskoster commented:
I understand this is subjective but I actually appreciate the `not-allowed` cursor in this case because it explicitly communicates that I am not allowed to perform the intended action. The `default` cursor is more ambiguous in that sense. 

I'd also suggest that using `pointer` for selectable items can be problematic because the table rows and cards often contain multiple interactive elements (which also exhibit a pointer cursor). Using `pointer` to indicate selection doesn't feel very reliable to me. When I see a pointer while hovering the title/thumbnail I'm unsure whether I'll be taken to the editor or not, despite the keypress. 

Another option to try would be; use `default` for selectable items and `not-allowed` for un-selectable items. With the hover styles it should be clear that the entire element is interactive which eliminates confusion about what a click might do. There's also some visual cohesion between `default` and `not-allowed` which might be helpful here too.

jasmussen commented:
> explicitly communicates that I am not allowed to perform the intended action

The problem here is that it isn't clear what _the intended action is_. There's nothing dangerous about you clicking a checkbox. But the default pointer suggests enough that it's not interactive.

This isn't a strong opinion on my part, my main concern is that it's surprising to see that cursor because you almost never see it, anywhere, to be honest. I don't recall any other website or app I've seen that cursor on.

jameskoster commented:
> isn't clear what the intended action is.

Doesn't the keypress make it clear? Cmd+click is a very common pattern for selection. 

It feels strange to me to see the `default` cursor on _unselectable_ items. That's the cursor you typically see for _selectable_, e.g. Google Drive.
